drivers/afe/ad4110/ad4110.c

```diff
@@ -530,70 +690,166 @@ int32_t ad4110_setup(struct ad4110_dev **device,
 
 	/* SPI */
 	ret = spi_init(&dev->spi_dev, &init_param.spi_init);
-
-	/* GPIO */
-	ret |= gpio_get(&dev->gpio_reset, &init_param.gpio_reset);
-
-	ret |= gpio_direction_output(dev->gpio_reset, GPIO_LOW);
-	mdelay(10);
-	ret |= gpio_set_value(dev->gpio_reset, GPIO_HIGH);
-	mdelay(10);
+	if (ret)
+		goto err_init;
```

Comment:
here you don't need to call `spi_remove`, if the initialization failed. The platform specific `spi_init` functions handle these deallocations already. Example:
https://github.com/analogdevicesinc/no-OS/blob/f0b2bd46dca3d41de76f431050126c89b964bd8a/drivers/platform/xilinx/xilinx_spi.c#L264